File: plugin/zeppelin/src/main/scala/jetbrains/zeppelin/idea/settings/plugin/RemoteRunSettingsState.scala
package jetbrains.zeppelin.idea.settings.plugin

import com.intellij.util.xmlb.annotations.{Attribute, Property}
import jetbrains.zeppelin.constants.ZeppelinConstants
import jetbrains.zeppelin.models.{SparkVersion, User}

case class RemoteRunSettingsState() {
  @Property(surroundWithTag = false) var zeppelinSettings: ZeppelinSettings = ZeppelinSettings()
}

class ZeppelinSettings {
  @SuppressWarnings(Array("FieldMayBeFinal"))
  @Attribute("defaultName")
  var defaultNotebookName: String = ZeppelinConstants.DEFAULT_NOTEBOOK_NAME

  @SuppressWarnings(Array("FieldMayBeFinal"))
  @Attribute("sparkVersion")
  var sparkVersion: String = SparkVersion.ZEPPELIN_DEFAULT_VERSION.toString

  @SuppressWarnings(Array("Address"))
  @Attribute("UriEnabled")
  var address: String = "localhost"

  @SuppressWarnings(Array("FieldMayBeFinal"))
  @Attribute("Login")
  var login: String = "admin"

  @SuppressWarnings(Array("FieldMayBeFinal"))
  @Attribute("Password")
  var password: String = "password"

  @SuppressWarnings(Array("FieldMayBeFinal"))
  @Attribute("Port")
  var port: Int = 8080

  @SuppressWarnings(Array("FieldMayBeFinal"))
  @Attribute("IsAnonymous")
  var isAnonymous: Boolean = false

  def setAddress(value: String): Unit = {
    address = value
  }

  def setIsAnonymous(value: Boolean): Unit = {
    isAnonymous = value
  }

  def setLogin(value: String): Unit = {
    login = value
  }

  def setPassword(value: String): Unit = {
    password = value
  }

  def setPort(value: Int): Unit = {
    port = value
  }

  def setSparkVersion(value: String): Unit = {
    sparkVersion = value
  }

  def setDefaultNotebookName(value: String): Unit = {
    defaultNotebookName = value
  }

  def user: Option[User] = {
    if (isAnonymous) return None
    Some(User(login, password))
  }

  def fullUrl: String = s"$address:$password"
}

object ZeppelinSettings {
  def apply(): ZeppelinSettings = new ZeppelinSettings()
}

trait ZeppelinHolder {
  def getZeppelinSettings: ZeppelinSettings

  def setZeppelinSettings(settings: ZeppelinSettings): Unit
}

File: zeppelin/implementation/src/org/intellij/plugin/zeppelin/idea/settings/plugin/RemoteRunSettingsState.kt
package org.intellij.plugin.zeppelin.idea.settings.plugin

import com.intellij.util.xmlb.annotations.Attribute
import com.intellij.util.xmlb.annotations.Property
import org.intellij.plugin.zeppelin.constants.ZeppelinConstants
import org.intellij.plugin.zeppelin.models.SparkVersion
import org.intellij.plugin.zeppelin.models.User

data class RemoteRunSettingsState(
        @Property(surroundWithTag = false)
        var zeppelinSettings: ZeppelinSettingsStorage = ZeppelinSettingsStorage())

class ZeppelinSettingsStorage {

    private var login: String = ZeppelinCredentialsManager.getLogin()
    private var password: String = ZeppelinCredentialsManager.getPlainPassword()

    @Attribute("defaultName")
    var defaultNotebookName: String = ZeppelinConstants.DEFAULT_NOTEBOOK_NAME

    @Attribute("sparkVersion")
    var sparkVersion: String = SparkVersion.ZEPPELIN_DEFAULT_VERSION.toString()

    @Attribute("host")
    var host: String = "localhost"

    @Attribute("port")
    var port: Int = 8080

    @Attribute("IsAnonymous")
    var isAnonymous: Boolean = false

    private fun setCredentials(login: String, password: String) {
        this.login = login
        this.password = password
    }

    val user: User?
        get() {
            if (isAnonymous) return null
            val login: String = ZeppelinCredentialsManager.getLogin()
            val password: String = ZeppelinCredentialsManager.getPlainPassword()
            return User(login, password)
        }

    var data: ZeppelinSettings
        get() = ZeppelinSettings(host, port, user, SparkVersion(sparkVersion), defaultNotebookName)
        set(value) {
            host = value.host
            port = value.port
            sparkVersion = value.sparkVersion.toString()
            defaultNotebookName = value.defaultNotebookName
            if (value.user != null) {
                isAnonymous = false
                setCredentials(login = value.user.name, password = value.user.password)
            } else {
                isAnonymous = true
            }
        }
}

data class ZeppelinSettings(val host: String,
                            val port: Int,
                            val user: User?,
                            val sparkVersion: SparkVersion,
                            val defaultNotebookName: String) {
    val isAnonymous
        get() = user == null

    val fullUrl: String
        get() = "$host:$port"

    val login
        get() = user?.name ?: ""

    val password
        get() = user?.password ?: ""
}

interface ZeppelinHolder {
    fun getZeppelinSettings(): ZeppelinSettings
    fun setZeppelinSettings(settings: ZeppelinSettings)
}